### PR TITLE
10765 offer option dublicate listview

### PR DESCRIPTION
**Pull request by hungnc89**

### CODE REVIEW

File: plugin/Record/RecordManagerDuplicateListViewEstate.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		$pRecordManagerReadListViewEstate = $this->_pContainer->get(RecordManagerReadListViewEstate::class);
+		$listViewRoot = $pRecordManagerReadListViewEstate->getRowById($id);
+
+		if (isset($listViewRoot) && $listViewRoot != null) {
+			$tableListViews = $prefix . self::TABLENAME_LIST_VIEW;
+
+			//append count name
+			$originalName = $listViewRoot['name'];
+			$newName = "{$originalName} - Copy";
+			$selectLikeOriginalName = "SELECT `name` FROM {$tableListViews} WHERE name LIKE '{$newName}%' ORDER BY listview_id DESC";
```
Comment (jayay) [inline plugin/Record/RecordManagerDuplicateListViewEstate.php:84]: SQL escaping is missing for `$newName`.

File: plugin/Record/RecordManagerDuplicateListViewEstate.php [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+				$tableListViews,
+				$newListView
+			);
+			$duplicateListViewId = $this->_pWPDB->insert_id;
+
+			if ($duplicateListViewId !== 0) {
+				//duplicate data related oo_plugin_fieldconfig table
+				$tableFieldConfig = $prefix . self::TABLENAME_FIELDCONFIG;
+				foreach ($listViewRoot['fields'] as $field) {
+					$selectFieldConfigByIdAndFieldName = "SELECT * FROM {$tableFieldConfig} WHERE listview_id='{$id}' AND fieldname ='{$field}'";
```
Comment (jayay) [inline plugin/Record/RecordManagerDuplicateListViewEstate.php:118]: SQL escaping should be added for `$id` as well

File: plugin/Record/RecordManagerDuplicateListViewEstate.php [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+							$newFieldConfigRow['hidden'] = $fieldConfigRow->hidden;
+							$newFieldConfigRow['availableOptions'] = $fieldConfigRow->availableOptions;
+							$this->_pWPDB->insert($tableFieldConfig, $newFieldConfigRow);
+						}
+					}
+				}
+
+				//duplicate data related oo_plugin_picturetypes table
+				$tablePictureTypes = $prefix . self::TABLENAME_PICTURETYPES;
+				$selectPictureTypesById = "SELECT * FROM {$tablePictureTypes} WHERE listview_id='{$id}'";
```
Comment (jayay) [inline plugin/Record/RecordManagerDuplicateListViewEstate.php:136]: Same here

File: plugin/Record/RecordManagerDuplicateListViewEstate.php [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+						$newPictureTypeRow = [];
+						$newPictureTypeRow['listview_id'] = $duplicateListViewId;
+						$newPictureTypeRow['picturetype'] = $pictureTypeRow->picturetype;
+						$this->_pWPDB->insert($tablePictureTypes, $newPictureTypeRow);
+					}
+				}
+
+				//duplicate data related oo_plugin_sortbyuservalues table
+				$tableSortByUserValues = $prefix . self::TABLENAME_SORTBYUSERVALUES;
+				$selectSortByUserValuesById = "SELECT * FROM {$tableSortByUserValues} WHERE listview_id='{$id}'";
```
Comment (jayay) [inline plugin/Record/RecordManagerDuplicateListViewEstate.php:149]: And again

**Review by jayay — CHANGES_REQUESTED** [review]
Please fix the SQL escaping here. Table names should be escaped too, if they come from a variable.

**Review by jayay — APPROVED** [review]
Thank you! Is this ready for QA then?

**Comment by nglelinh** [comment]
Yes, the pull request is ready for QA

File: plugin/Gui/AdminPageEstate.php [hunk 5]
```diff
@@ -231,9 +232,24 @@ public function preOutput()
 			}
 			return $redirectTo;
 		};
-
 		add_filter('handle_bulk_actions-onoffice_page_onoffice-estates', $pClosureDeleteEstate, 10, 3);
 
+		$pClosureDuplicateEstate = function (string $redirectTo, Table\WP\ListTable $pTable)
+		use ($pDI): string {
+			if (in_array($pTable->current_action(), ['duplicate', 'bulk_duplicate'])) {
+				if (!(isset($_GET['listVewId']))) {
```
Comment (jayay) [inline plugin/Gui/AdminPageEstate.php:240]: Noticed this last-minute: the IDs should be a parameter of the function like in line 219.
`check_admin_referer` must be used to check if the action was triggered by someone logged in with an admin account.

**Review by jayay — CHANGES_REQUESTED** [review]
I found one more thing related to security. Maybe this should be checked.

File: plugin/Gui/AdminPageEstate.php (hunk 5 above)
Comment (hungnc89) [reply plugin/Gui/AdminPageEstate.php:240]: yes, I will check and fix it.

Comment (jayay) [reply plugin/Gui/AdminPageEstate.php:241]: Thanks!

**Review by jayay — APPROVED** [review]

**Comment by michaelarnold-dev** [comment]
@nglelinh : There have been some conflicts which restricts us merging the branch. Could you please solve the conflicts?

**Comment by nglelinh** [comment]
@michaelarnold-dev please help to continue, thank you

**Pull request merged**